[SaferCPP] Address issues in WebCore/css/parser

roberto-apple opened the pull request:
#### ec18ba9c09e8820d30ed23177cbdae52d9521d64
<pre>
[SaferCPP] Address issues in WebCore/css/parser
<a href="https://bugs.webkit.org/show_bug.cgi?id=291182">https://bugs.webkit.org/show_bug.cgi?id=291182</a>
<a href="https://rdar.apple.com/148718797">rdar://148718797</a>

Reviewed by Chris Dumez.

Address SaferCPP issues in WebCore/css/parser.

* Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/css/RectBase.h:
(WebCore::RectBase::protectedTop const):
(WebCore::RectBase::protectedRight const):
(WebCore::RectBase::protectedBottom const):
(WebCore::RectBase::protectedLeft const):
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeRuleList):
* Source/WebCore/css/parser/CSSParserImpl.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::isNumber):
(WebCore::CSSPropertyParser::consumeAnimationRangeShorthand):
* Source/WebCore/css/parser/CSSPropertyParser.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Grid.cpp:
(WebCore::CSSPropertyParserHelpers::isGridTrackFixedSized):
* Source/WebCore/css/parser/SizesCalcParser.h:

Canonical link: <a href="https://commits.webkit.org/293502@main">https://commits.webkit.org/293502@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/928114f7a1516cd8f8c13c4285ba228febe20dfe

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99024 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18665 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8909 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104153 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49617 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101069 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18957 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27111 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75385 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32510 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102028 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14457 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89436 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55746 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14213 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7408 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48993 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84151 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7491 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106519 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26121 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19064 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84346 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26516 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85633 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83872 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21283 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28523 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6190 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/19858 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26074 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31260 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25899 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29230 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27468 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Grid.cpp:
```diff
@@ -215,13 +215,15 @@ static bool isGridTrackFixedSized(const CSSPrimitiveValue& primitiveValue)
 {
-    if (auto* primitiveValue = dynamicDowncast<CSSPrimitiveValue>(value))
+    RefPtr primitiveValue = dynamicDowncast<CSSPrimitiveValue>(value);
+    if (primitiveValue)
```
```suggestion
    if (RefPtr primitiveValue = dynamicDowncast<CSSPrimitiveValue>(value))
```

roberto-apple replied:
Updated

nt1m commented on Source/WebCore/css/parser/SizesCalcParser.cpp:
```diff
@@ -102,7 +102,8 @@ bool SizesCalcParser::appendLength(const CSSParserToken& token)
     SizesCalcValue value;
-    double result = SizesAttributeParser::computeLength(token.numericValue(), *lengthUnit, m_document);
+    const Ref<Document> document = m_document;
+    double result = SizesAttributeParser::computeLength(token.numericValue(), *lengthUnit, document);
```
Is this needed if m_document is const ?

cdumez replied:
@nt1m is right, you shouldn't need to.

roberto-apple replied:
This was the issue I was addressing: https://build.webkit.org/results/Apple-Sequoia-Safer-CPP-Checks/293460%40main%20%282949%29/scan-build-output/StaticAnalyzer/WebCore/StaticAnalyzerReports/report-SizesCalcParser.cpp-appendLength-5-9c4c48.html#EndPath

Could it have been a false positive? With respective to the above comment about using `Ref<const Document>` has the data member, maybe that alone will address it.

cdumez replied:
It was not a false positive but it predates you marking the data member as `const` in your own PR.

cdumez replied:
This fix you made in your PR:
```
const Ref<Document> m_document;
```

Addressed this warning already.

cdumez commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -238,12 +238,12 @@ RefPtr<StyleRuleNestedDeclarations> CSSParserImpl::parseNestedDeclarations(const
 {
     CSSParserImpl parser(context, string, &styleSheet, nullptr);
-    bool firstRuleValid = parser.consumeRuleList(parser.tokenizer()->tokenRange(), RuleList::TopLevel, [&](Ref<StyleRuleBase> rule) {
+    bool firstRuleValid = parser.consumeRuleList(parser.tokenizer()->tokenRange(), RuleList::TopLevel, [&context, styleSheet = Ref { styleSheet }](Ref<StyleRuleBase> rule) {
```
`consumeRuleList()` calls its lambda synchronously so we do not need any changes here. I think we just need to annotate the lambda passed to `consumeRuleList()` with `NOESCAPE`

roberto-apple replied:
Updated consumeRuleList to be NOESCAPE

cdumez commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -315,10 +315,10 @@ void CSSParserImpl::parseStyleSheetForInspector(const String& string, const CSSP
     CSSParserObserverWrapper wrapper(observer);
     CSSParserImpl parser(context, string, &styleSheet, &wrapper);
-    bool firstRuleValid = parser.consumeRuleList(parser.tokenizer()->tokenRange(), RuleList::TopLevel, [&styleSheet](Ref<StyleRuleBase> rule) {
+    bool firstRuleValid = parser.consumeRuleList(parser.tokenizer()->tokenRange(), RuleList::TopLevel, [styleSheet = Ref { styleSheet }](Ref<StyleRuleBase> rule) {
```
Should not be needed, the previous code was safe. Same issue as above.

roberto-apple replied:
Updated consumeRuleList to be NOESCAPE

cdumez commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -3268,7 +3268,8 @@ bool CSSPropertyParser::consumeAnimationRangeShorthand(CSS::PropertyParserState&
             else {
                 RefPtr startPair = downcast<CSSValuePair>(start);
-                end = rangeEndValueForStartValue(startPair->first());
+                Ref first = startPair->first();
```
You should call `protectedFirst()` below instead. It even already exists.

roberto-apple replied:
Updated

cdumez commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -1525,10 +1525,10 @@ static bool isNumber(const CSSValue& value, double number, CSSUnitType type)
-        && isNumber(quad.right(), number, type)
-        && isNumber(quad.bottom(), number, type)
-        && isNumber(quad.left(), number, type);
+    return isNumber(Ref { quad.top() }, number, type)
```
Ideally we'd add ` protectedTop()` function to RectBase. Same for the 3 below.

roberto-apple replied:
Added protected accessors to RectBase

cdumez commented on Source/WebCore/css/parser/SizesCalcParser.cpp:
```diff
@@ -38,7 +38,7 @@ namespace WebCore {
 SizesCalcParser::SizesCalcParser(CSSParserTokenRange range, const Document& document)
     : m_result(0)
-    , m_document(document)
+    , m_document(const_cast<Document&>(document))
```
Can we use a `Ref<const Document>` instead for the data member and avoid the ugly const casting?

roberto-apple replied:
Updated member variable

cdumez reviewed (approved):
LGTM

roberto-apple committed:
[SaferCPP] Address issues in WebCore/css/parser
https://bugs.webkit.org/show_bug.cgi?id=291182
rdar://148718797

Reviewed by Chris Dumez.

Address SaferCPP issues in WebCore/css/parser.

* Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/css/RectBase.h:
(WebCore::RectBase::protectedTop const):
(WebCore::RectBase::protectedRight const):
(WebCore::RectBase::protectedBottom const):
(WebCore::RectBase::protectedLeft const):
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeRuleList):
* Source/WebCore/css/parser/CSSParserImpl.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::isNumber):
(WebCore::CSSPropertyParser::consumeAnimationRangeShorthand):
* Source/WebCore/css/parser/CSSPropertyParser.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Grid.cpp:
(WebCore::CSSPropertyParserHelpers::isGridTrackFixedSized):
* Source/WebCore/css/parser/SizesCalcParser.h:

Canonical link: https://commits.webkit.org/293502@main
